staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -0,0 +1,211 @@
+	requestDigest utilflowcontrol.RequestDigest,
+	noteFn func(fs *fctypesv1a1.FlowSchema, pl *fctypesv1a1.PriorityLevelConfiguration),
+	queueNoteFn fq.QueueNoteFn,
+	execFn func(),
+) {
+	if t.mockDecision == decisionError {
+		panic("Handle should not be invoked")
+	}
+	noteFn(bootstrap.SuggestedFlowSchemaGlobalDefault, bootstrap.SuggestedPriorityLevelConfigurationGlobalDefault)
+	switch t.mockDecision {
+	case decisionExempt:
+		execFn()
+	case decisionExecute:
+		queueNoteFn(true)
+		queueNoteFn(false)
+		execFn()
+	case decisionCancelWait:
+		queueNoteFn(true)
+		queueNoteFn(false)
+	case decisionReject:
+		return
+	}
+}
+
+func (t fakeApfFilter) Run(stopCh <-chan struct{}) error {
+	return nil
+}
+
+func (t fakeApfFilter) Install(c *mux.PathRecorderMux) {
+}
+
+func newApfServer(decision int) *httptest.Server {
+	requestInfoFactory := &apirequest.RequestInfoFactory{APIPrefixes: sets.NewString("apis", "api"), GrouplessAPIPrefixes: sets.NewString("api")}
+	longRunningRequestCheck := BasicLongRunningRequestCheck(sets.NewString("watch"), sets.NewString("proxy"))
+
+	apfHandler := WithPriorityAndFairness(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+	}), longRunningRequestCheck, fakeApfFilter{mockDecision: decision})
+	handler := apifilters.WithRequestInfo(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		r = r.WithContext(apirequest.WithUser(r.Context(), &user.DefaultInfo{
+			Groups: []string{user.AllUnauthenticated},
+		}))
+		apfHandler.ServeHTTP(w, r)
+	}), requestInfoFactory)
+
+	apfServer := httptest.NewServer(handler)
+	return apfServer
+}
+
+func TestApfSkipLongRunningRequest(t *testing.T) {
+	epmetrics.Register()
+
+	server := newApfServer(decisionError)
+	defer server.Close()
+
+	if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces?watch=true", server.URL), http.StatusOK); err != nil {
+		// request should not be rejected
+		t.Error(err)
+	}
+
+	checkForExpectedMetrics(t, []string{
+		"apiserver_current_inflight_requests",
+	})
+}
+
+func TestApfRejectRequest(t *testing.T) {
+	epmetrics.Register()
+
+	server := newApfServer(decisionReject)
+	defer server.Close()
+
+	if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusTooManyRequests); err != nil {
+		t.Error(err)
+	}
+
+	// wait metrics to be collected
+	time.Sleep(time.Millisecond * 1500)
+
+	checkForExpectedMetrics(t, []string{
+		"apiserver_request_terminations_total",
+		"apiserver_dropped_requests_total",
+	})
+}
+
+func TestApfExemptRequest(t *testing.T) {
+	epmetrics.Register()
+	fcmetrics.Register()
+
+	// wait ReadWriteConcurrencyObserverPairGenerator to be registered
+	time.Sleep(time.Millisecond * 50)
+
+	server := newApfServer(decisionExempt)
+	defer server.Close()
+
+	if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusOK); err != nil {
+		t.Error(err)
+	}
+
+	// wait metrics to be collected
+	time.Sleep(time.Millisecond * 1500)
```

Comment:
On second throught, the problem here is relying on the CPU scheduling to get asynchronous work done by any particular deadline.  It would be more robust here to have a loop that sleeps for a little while (one second would be fine) and then looks for all the expected metrics, terminating the loop successfully if they are all found.  Put a limit on the number of iterations, and complain about the metrics not found in the last iteration.